### ISSUE TITLE
performance issue with _getkey in cidict.py

### ISSUE THREAD

**Issue by pmellier** [cannatag/ldap3#109]
Hi,

I made some performance comparisons between AD and OpenLDAP using ldap3, and I found that AD was very very slow. 

After some investigations with a profiling tools, I found that a lot of time was spent in _getkey from cidict.py. A simple BIND takes 5s, cidict.py was called 5491 time (the same test with OpenLdap reports 0.09s and 662 calls).

After updating the code by adding a _insensitive_key_map dict to save the correspondance between key.lower() and key, and removing the loop, the performance was 0.67s for the BIND in AD and 0.07s for OpenLDAP.

Best regards,

Pierre Mellier

**Comment by cannatag** [comment]
Thanks a lot Pierre, I never tested the cud icy performance. The loop is obviously useless in _getkey. I will fix it in the next release.

Bye,
Giovanni

**Comment by cannatag** [comment]
Can you check code in dev? I've implemented a keymap dict as you suggested.

**Comment by pmellier** [comment]
Giovanni,

I check your code. My only remark concerns the __delitem__ function. Is it not better to move the last function line inside the if ?

—old:
def __delitem__(self, key):
        ci_key = self._ci_key(key)
        if ci_key in self._case_insensitive_keymap: # updates existing value
        del self._store[self._case_insensitive_keymap[ci_key]]
        else:
        del self._store[key]
        del self._case_insensitive_keymap[self._ci_key(key)]


—new:
def __delitem__(self, key):
        ci_key = self._ci_key(key)
        if ci_key in self._case_insensitive_keymap: # updates existing value
        del self._store[self._case_insensitive_keymap[ci_key]] del self._case_insensitive_keymap[ci_key]
        else:
        del self._store[key]

I think you could also remove the _case_insensitive_types variable ?

Best regards, any many thanks for your very fast answer !

Pierre

Le 19 sept. 2015 à 17:10, Giovanni Cannata <notifications@github.com<mailto:notifications@github.com>> a écrit :


Can you check code in dev? I've implemented a keymap dict as you suggested.

**Comment by cannatag** [comment]
Hi Pierre,
I think the __delitem__ method could be still more simple:

```   
    def __delitem__(self, key):
        ci_key = self._ci_key(key)        
        del self._store[self._case_insensitive_keymap[ci_key]]
        del self._case_insensitive_keymap[self._ci_key(key)]
```

it should cover all cases (sensitive and insensitive). The latest line is needed to remove the key from the mapping.

You're right about  _case_insensitive_types, it is now useless because I check the presence of the lower() method to decide if the key can be cast to lower case.

You can check the updated code in dev.

Bye,
Giovanni

**Comment by pmellier** [comment]
Giovanni,

I agree your changes. Perhaps you could also use the ci_key variable in the last line for clarity.

Best regards,

Pierre

Le 22 sept. 2015 à 17:17, Giovanni Cannata <notifications@github.com<mailto:notifications@github.com>> a écrit :


Hi Pierre,
I think the delitem method could be still more simple:

    def __delitem__(self, key):
        ci_key = self._ci_key(key)
        del self._store[self._case_insensitive_keymap[ci_key]]
        del self._case_insensitive_keymap[self._ci_key(key)]


it should cover all cases (sensitive and insensitive). The latest line is needed to remove the key from the mapping.

You're right about _case_insensitive_types, it is now useless because I check the presence of the lower() method to decide if the key can be cast to lower case.

You can check the updated code in dev.

Bye,
Giovanni

**Comment by cannatag** [comment]
right. you're a perfect code reviewer! 

thanks a lot.
Giovanni

**Comment by cannatag** [comment]
released in 0.9.9.1
